ADR/ADR040-make-container-image-tags-mutable.md

```diff
@@ -0,0 +1,45 @@
+# ADR040: Make container image tags mutable
+
+Date: 2025-02-12
+
+## Status
+
+> Accepted or Rejected
+>
+> Valid ADR statuses are: "Proposed", "Accepted", "Rejected", "Deprecated", "Superseded by ADRzzz", but since we use the GitHub Pull request mechanism to approve ADRs, "Proposed" isn't applicable.
+
+## Context
+
+At the time of writing, we use [Amazon ECR](https://aws.amazon.com/ecr/) to host our container image repositories. 
+
+Each container image can have any number of tags applied to it, consistent with [the Open Container Initiative spec](https://github.com/opencontainers/distribution-spec/blob/main/spec.md). No two container images can have the same tag, and attempting to apply the same tag to a second image will result in an error. This is called tag immutability, and is [a feature of AWS ECR](https://docs.aws.amazon.com/AmazonECR/latest/userguide/image-tag-mutability.html) that we have enabled.
+
+The alternative to tag immutability is tag mutability. With tag mutability enabled, when a tag is applied to a second container image, the tag is removed from the first one before being applied to the second. It is, in effect, a move operation. In most cases this is an undesirable property of an image repository, because it allows for situations where downloading a container image with a given tag on two occasions can result in different images; you can never be absolutely sure that you got the same bytes twice.
+
+We are implementing review applications for `forms-admin` (see [ADR037](./ADR037-implement-review-apps.md)). In our implementation we want to deploy the version of `forms-admin` under review alongside the latest production version of `forms-api`. Deploying the latest production version means two things:
+
+1. we don't need to constantly update the code to change the version being deployed
+2. we need a mechanism for consistently identifying the latest production version
+
+The `latest` tag is conventionally the tag used to denote the most recent version of a container image, or otherwise the latest production-ready image. We see no reason not to follow convention, and have begun [tagging the newest production-ready images](https://github.com/alphagov/forms-deploy/pull/1346). 
```

Comment:
Nit: How are we adding the "latest" tag? In a lot of cases "latest" means most recently pushed image to the registry, rather than most "recent version".

Reply:
The PR linked there (`forms-deploy` #1346) has the implementation. But in short, we have a flag set in the production container deploying pipelines which add the latest tag at the same time they add the tag which means "this ran fine in production". None of those pipelines can have two run happens simultaneously; that is to say two deployments of `forms-admin` can't be happening concurrently, but `forms-admin` and `forms-api` can be deployed at the same time.